### DISCUSSION TITLE
Validation of the generated template

### DISCUSSION THREAD

**Discussion by jeremych1000** [hairyhenderson/gomplate#1681] [Q&A] [answered]
Hello! Thanks for making this tool. I'm using it atm to template out some YAML files, and want to do some validation to make sure that the values are within a certain range, and are of a certain type.

Which is the recommended way? I want to do all the validation inside the template, and don't want to have to invoke a downstream tool to parse the generated YAML file.

I tried a mixture of the below, but could never quite get it working correctly. What am I doing wrong?

Assuming an input JSON data source of 
```json
{ 
  "tenant": {
      "name": "foobar",
      "account": 123123,
  }
}
```

I want to verify that the tenant.name is a string, and that the account is an integer, and between 0-999999.

I tried things like:
`{{ $tenant.name | test.IsKind "string" }}` <-- this just prints true/false in my template
`{{ $tenant.name | test.Assert ( eq $tenant.name | test.Kind "string ) }}` <-- failed saying it got a bool instead
etc. etc.

**Comment by hairyhenderson** [comment, marked as answer]
Hi @jeremych1000, I've transferred this to a Discussion since this wasn't a bug report or feature request.

Here's a sample template that will do what I think you're looking for:

```go
{{ $nameKind := test.Kind .tenant.name -}}
{{ $acctKind := test.Kind .tenant.account -}}
{{ eq "string" $nameKind | test.Assert (printf "name must be a string, but was a %s" $nameKind) -}}
{{ eq "int" $acctKind | test.Assert (printf "account must be an int, but was a %s" $acctKind) -}}
```

Given an (invalid) input of:

```json
{ 
  "tenant": {
      "name": "foobar",
      "account": true
  }
}
```

This will fail as follows:

```console
$ gomplate -c .=in.json -f t.tmpl
09:21:12 ERR error="renderTemplate: failed to render template t.tmpl: template: t.tmpl:4:28: executing \"t.tmpl\" at <test.Assert>: error calling Assert: assertion failed: account must be an int, but was a bool"
```

The main issue here is that the arguments to `test.Assert` were reversed in your case. With `test.Assert, the last arg (or piped arg) needs to be a boolean. If two args are provided, the first argument is an optional error message.

Have a look at the [`test.Assert` docs](https://docs.gomplate.ca/functions/test/#test-assert) for full details.

Hope that helps!

**Reply by jeremych1000** [reply]
Perfect, thanks this worked. 

(has been working since March actually, so closing for completeness :))
